codegen/layout_tool.py

```diff
@@ -0,0 +1,314 @@
+                int(x - class_ref.border_offsets // class_ref.scalar), 
+                int(y - class_ref.border_offsets // class_ref.scalar)
+            )
+            
+            collision_map = {}
+            for e in class_ref.layout_info.keys():
+                collision_map[e] = False
+
+            for ent_name, data in class_ref.layout_info.items():
+                    for o_name, odata in class_ref.layout_info.items():
+                            if ent_name + "__border" != o_name and ent_name != o_name + "__border":
+                                if ent_name != o_name:
+                                    if check_collision(data, odata):
+                                        collision_map[ent_name] = True
+                                        collision_map[o_name] = True
+            
+            for ent_name, data in class_ref.layout_info.items():
+                if "__border" in ent_name:
+                    if collision_map[ent_name]:
+                        class_ref.layout_info[ent_name]["design_color"] = class_ref.bad_color
+                        if layout_state != "bad":
+                            layout_state = "warning"
+                        # if there is a bad color on border, we want to add warning to main
+                        e = ent_name.replace("__border", "")
+                        # but only if the color isn't already bad
+                        if (class_ref.layout_info[e]["design_color"] == class_ref.good_color).all():
+                            class_ref.layout_info[e]["design_color"] = class_ref.warning_color
+                            if layout_state != "bad":
+                                layout_state = "warning"
+                    else:
+                        class_ref.layout_info[ent_name]["design_color"] = class_ref.warning_color
+                else:
+                    if collision_map[ent_name]:
+                        class_ref.layout_info[ent_name]["design_color"] = class_ref.bad_color
+                        layout_state = "bad"
+                    else:
+                        class_ref.layout_info[ent_name]["design_color"] = class_ref.good_color
+
+            class_ref.update_image()
+    pass
+
+
+class LayoutTool():
+    def __init__(self, collection, downscale):
+        logging.info(f"launching layout tool. Found {len(collection.projects)} designs in collection")
+        self.callback = None
+        self.scalar = downscale
+        self.area_width = int(collection.config['configuration']['user_area_width'] // self.scalar)
+        self.area_height = int(collection.config['configuration']['user_area_height'] // self.scalar)
+        logging.info(f"layout tool canvas = {self.area_width}x{self.area_height}, using downscale = {self.scalar}")
+        logging.info(f"pass --layout-tool-downscale=x if this does not fit on your screen")
+ 
+        self.pointerbound = None
+
+        # values for the color scheme, these should probably be in config?
```

Comment:
Not sure how much configuration we want here. right now I've stolen the discord theme, but It might make sense to keep it in config file for accessibility sake

Reply:
not important for me